Add style controls for order shipping and billing address blocks

tarhi-saad opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
In this PR, we added style controls (i.e., Typography, colors, border styles) on the settings sidebar of the `Shipping Address` and `Billing Address` blocks.

<img width="272" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/14235870/a1ce8e83-ea9b-4723-bfa2-624fa0423ab9">
<!-- Reference any related issues or PRs here -->

Fixes #10120

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

#### Other Checks

- [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Testing

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Go to the Order Confirmation on the editor side
2. Open the `Settings` sidebar
3. Select the `Billing Address` or `Shipping Address` inner block
4. Ensure the `Color`, `Typography`, and `Border` style controls are available
5. Change the style of the Block, and save.
6. Reload the page. Ensure the changes are still applied
7. Place an order on the client side. Ensure we have the same style on the Order Confirmation page.

* [x] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested (ie: it makes changes to tests, coding standards, docblocks, etc.). -->
* [ ] Should be tested by the development team exclusively <!-- Check this box if this PR should be tested by the development team exclusively (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

tarhi-saad committed:
Add style controls for Shipping and Billing Blocks

nielslange commented on assets/js/blocks/order-confirmation/shipping-address/style.scss:
```diff
@@ -0,0 +1,9 @@
+.wc-block-order-confirmation-shipping-address {
+	address {
+		width: 100% !important;
```
Why are we using `width: 100% !important;` instead of  `width: 100%;` here? I'm asking as `!important` should only be used as last option. In this case, even when removing `width: 100% !important;` completely, I do not see a change on the site.

nielslange commented on assets/js/blocks/order-confirmation/shipping-address/style.scss:
```diff
@@ -0,0 +1,9 @@
+		padding: $gap-large;
+		display: block;
+		box-sizing: border-box;
+		border: 1px solid currentColor;
```
The definition `border: 1px solid currentColor;` seems to cause an extra inner border:

<img width="1079" alt="Screenshot 2023-08-17 at 11 31 05" src="https://github.com/woocommerce/woocommerce-blocks/assets/3323310/12a35edd-80c2-44fe-b474-5ba9c2fc8597">

nielslange commented on assets/js/blocks/order-confirmation/status/style.scss:
```diff
@@ -0,0 +1,16 @@
+		}
+		a.has-link-color {
+			color: inherit;
+		}
```
Shouldn't we fully utilise SCSS in the following case:

```suggestion
		a {
			margin-right: $gap;		
			
			&.has-link-color {
				color: inherit;
			}
		}
```

nielslange commented on assets/js/blocks/order-confirmation/summary/edit.tsx:
```diff
@@ -0,0 +1,41 @@
+	// Remove style because its handled by the server-side render.
+	const { style, ...blockProps } = useBlockProps( {
+		className: 'wc-block-order-confirmation-summary',
+	} );
```
Can you explain this part? The comment says "Remove style [...]", but the code looks like we're adding a class.

nielslange commented on assets/js/blocks/order-confirmation/downloads/style.scss:
```diff
@@ -0,0 +1,27 @@
+		td {
+			border-style: solid;
+			border-color: currentColor;
+			border-width: 0 0 1px 0;
```
We could use the shorthand syntax here:

```suggestion
			border: 0 0 1px 0 solid currentColor;
```

nielslange commented on assets/js/blocks/order-confirmation/billing-address/style.scss:
```diff
@@ -0,0 +1,9 @@
+.wc-block-order-confirmation-billing-address {
+	address {
+		width: 100% !important;
```
Why are we using `width: 100% !important;` instead of  `width: 100%;` here? I'm asking as `!important` should only be used as last option. In this case, even when removing `width: 100% !important;` completely, I do not see a change on the site.

nielslange commented on assets/js/blocks/order-confirmation/billing-address/style.scss:
```diff
@@ -0,0 +1,9 @@
+		padding: $gap-large;
+		display: block;
+		box-sizing: border-box;
+		border: 1px solid currentColor;
```
The definition `border: 1px solid currentColor;` seems to cause an extra inner border:

<img width="1079" alt="Screenshot 2023-08-17 at 11 31 05" src="https://github.com/woocommerce/woocommerce-blocks/assets/3323310/12a35edd-80c2-44fe-b474-5ba9c2fc8597">

nielslange commented on src/BlockTypes/OrderConfirmation/AbstractOrderConfirmationBlock.php:
```diff
@@ -0,0 +1,207 @@
+		// If the user is logged out, check the order key for validity.
+		if ( $this->allow_guest_checkout() && $this->is_current_customer_order( $order ) && $this->has_valid_order_key( $order ) ) {
+			return $this->order_matches_session( $order ) ? 'full' : 'limited';
+		}
```
What are your thoughts on placing the comment outside the if-statement so that it's consistent with the comment of the next if-statement?

```suggestion
		// If logged in, check the user owns the order.
		if ( is_user_logged_in() ) {
			return $this->is_current_customer_order( $order ) ? 'full' : false;
		}

		// If the user is logged out, check the order key for validity.
		if ( $this->allow_guest_checkout() && $this->is_current_customer_order( $order ) && $this->has_valid_order_key( $order ) ) {
			return $this->order_matches_session( $order ) ? 'full' : 'limited';
		}
```

nielslange commented on src/BlockTypes/OrderConfirmation/AbstractOrderConfirmationBlock.php:
```diff
@@ -0,0 +1,207 @@
+	 * Get the frontend script handle for this block type.
+	 *
+	 * @param string $key Data to get, or default to everything.
+	 */
```
The return type seems to be missing here:

```suggestion
	/**
	 * Get the frontend script handle for this block type.
	 *
	 * @param string $key Data to get, or default to everything.
	 * @return null
	 */
```

nielslange commented on src/BlockTypes/OrderConfirmation/AbstractOrderConfirmationBlock.php:
```diff
@@ -0,0 +1,207 @@
+	/**
+	 * Get a fake order for previews.
+	 *
+	 * @return \WC_Order Fake order.
```
There's a fullstop behind the return type. This is only required for comments, not for return values.

```suggestion
	 * @return \WC_Order Fake order
```

nielslange commented on src/BlockTypes/OrderConfirmation/Status.php:
```diff
@@ -0,0 +1,114 @@
+	protected function render_content( $order, $permission = false, $attributes = [] ) {
+		if ( ! $permission ) {
+			// phpcs:ignore WooCommerce.Commenting.CommentHooks.MissingHookComment
+			return '<p>' . wp_kses_post( apply_filters( 'woocommerce_thankyou_order_received_text', esc_html__( 'Thank you. Your order has been received.', 'woo-gutenberg-products-block' ), null ) ) . '</p>';
```
What are your thoughts on formatting this section similar to the next one for readability-reasons:

```suggestion
			return '<p>' . wp_kses_post( 
				apply_filters( 
					'woocommerce_thankyou_order_received_text', 
					esc_html__( 'Thank you. Your order has been received.', 'woo-gutenberg-products-block' ), 
					null 
				)
			 ) . '</p>';
```

nielslange commented on src/BlockTypes/OrderConfirmation/Status.php:
```diff
@@ -0,0 +1,114 @@
+	 * @return string
+	 */
+	protected function render_content_fallback() {
+		// phpcs:ignore WooCommerce.Commenting.CommentHooks.MissingHookComment
```
Can you explain the purpose of `// phpcs:ignore WooCommerce.Commenting.CommentHooks.MissingHookComment` in this case? It seems there's no hook that needs commenting.

nielslange commented on src/BlockTypes/OrderConfirmation/Summary.php:
```diff
@@ -0,0 +1,105 @@
+	 * @return string
+	 */
+	protected function render_summary_row( $name, $value ) {
+		return $value ? '<li class="wc-block-order-confirmation-summary-list-item"><span class="wc-block-order-confirmation-summary-list-item__key">' . esc_html( $name ) . '</span> <span class="wc-block-order-confirmation-summary-list-item__value">' . wp_kses_post( $value ) . '</span></li>' : '';
```
What are your thoughts on splitting the code into multiple lines to increase the readability:

```suggestion
		    if ( $value ) {
        return '<li class="wc-block-order-confirmation-summary-list-item">'
            . '<span class="wc-block-order-confirmation-summary-list-item__key">' 
            . esc_html( $name ) 
            . '</span> '
            . '<span class="wc-block-order-confirmation-summary-list-item__value">' 
            . wp_kses_post( $value ) 
            . '</span>'
            . '</li>';
    } else {
        return '';
    }
```

nielslange reviewed (commented):
Thanks for working on this, @tarhi-saad. Below, I left a few comments.

Apart from the comments, I wonder why this PR is so big. I wonder because other PRs that added style controls, only touched a handful of files, e.g. https://github.com/woocommerce/woocommerce-blocks/pull/10185/files.

mikejolley commented:
@tarhi-saad please change the base from trunk to the feature branch. That will cause the diff to be smaller (cc @nielslange)

tarhi-saad commented:
@nielslange, @mikejolley, I changed the base branch! Now you can see only my changes!

tarhi-saad committed:
Remove address tag borders

The wrapper border will be the one being updated. So, having another
inner border is a styling issue!

mikejolley commented:
@tarhi-saad the default has padding but no border now:

![Screenshot 2023-08-17 at 17 02 32](https://github.com/woocommerce/woocommerce-blocks/assets/90977/9efb6b60-03f2-43ac-bc97-6482be02a3f6)

I think it would be good to add spacing controls (to control padding) if we're leaving the border off. I wonder if its also possible to set defaults, so we have the border and padding (as originally designed) by default?

tarhi-saad committed:
Add spacing and fix border issue

tarhi-saad commented:
@mikejolley, would you mind reviewing the adjustments? I've moved the `padding` and `border` styles to the top-level HTML element rather than the `address` tag. This ensures that any modifications via the style controls will override the initial styles!

nielslange reviewed (commented):
The changes work as expected and the requirements of #10120 are met. 

However, I noticed that the PR introduces various TS errors, which might be coming from the underlying branch [`feature/blockify-order-confirmation`](https://github.com/woocommerce/woocommerce-blocks/tree/feature/blockify-order-confirmation). In addition, all [E2E tests](https://github.com/woocommerce/woocommerce-blocks/actions/runs/5897898474) and the [JS Unit Tests](https://github.com/woocommerce/woocommerce-blocks/actions/runs/5897898483) are failing. Before merging this PR, we should ensure that all tests are passing.

tarhi-saad commented:
@nielslange, thank you for the review! 🙌 

> However, I noticed that the PR introduces various TS errors, which might be coming from the underlying branch [feature/blockify-order-confirmation](https://github.com/woocommerce/woocommerce-blocks/tree/feature/blockify-order-confirmation). In addition, all [E2E tests](https://github.com/woocommerce/woocommerce-blocks/actions/runs/5897898474) and the [JS Unit Tests](https://github.com/woocommerce/woocommerce-blocks/actions/runs/5897898483) are failing. Before merging this PR, we should ensure that all tests are passing.

This PR is set to merge into the feature branch. We can proceed with the merge even if the E2E test fails, as we intend to address those failures within the feature branch itself. (See @mikejolley's status update: pdFofs-17q-p2#comment-1590)

nielslange commented:
> This PR is set to merge into the feature branch. We can proceed with the merge even if the E2E test fails, as we intend to address those failures within the feature branch itself. (See @mikejolley's status update: pdFofs-17q-p2#comment-1590)

Sounds good, as long as we ensure all tests are green before merging the featured branch into trunk.
